fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/ConsensusNode.java

```diff
@@ -1,94 +1,90 @@
+  private final Set<ElectInstance> electInstances;
 
   public ConsensusNode(PublicKey publicKey) {
     this.publicKey = publicKey;
-    this.consensuses = new ArrayList<>();
+    this.electInstances = new HashSet<>();
     this.acceptedMessageIds = new HashSet<>();
   }
 
   public PublicKey getPublicKey() {
     return publicKey;
   }
 
+  /**
+   * Get a set of messageId of Elect that this node has accepted.
+   *
+   * @return the set of accepted Elect messageId
+   */
   public Set<MessageID> getAcceptedMessageIds() {
     return Collections.unmodifiableSet(acceptedMessageIds);
   }
 
-  public Optional<Consensus> getLastConsensus(String instanceId) {
-    // get the consensus for the given instanceId, with the largest creation time
-    return consensuses.stream()
-        .filter(consensus -> consensus.getId().equals(instanceId))
-        .max(Comparator.comparingLong(Consensus::getCreation));
+  /**
+   * Get the latest ElectInstance that this node has created for the given instanceId.
+   *
+   * @param instanceId the id of the consensus
+   * @return an Optional ElectInstance
+   */
+  public Optional<ElectInstance> getLastElectInstance(String instanceId) {
+    // get the latest ElectInstance for the given instanceId
+    return electInstances.stream()
+        .filter(elect -> elect.getInstanceId().equals(instanceId))
+        .max(Comparator.comparingLong(ElectInstance::getCreation));
   }
 
+  /**
+   * Get the state of this node for the latest ElectInstance that this node has created for the
+   * given instanceId. If empty, it will return WAITING.
+   *
+   * @param instanceId the id of the consensus
+   * @return the current state
+   */
   public State getState(String instanceId) {
-    Optional<Consensus> lastConsensus = getLastConsensus(instanceId);
-    if (lastConsensus.isPresent()) {
-      Consensus consensus = lastConsensus.get();
-      if (consensus.isFailed()) {
-        return State.FAILED;
-      } else if (consensus.isAccepted()) {
-        return State.ACCEPTED;
-      } else {
-        return State.STARTING;
-      }
-    } else {
-      return State.WAITING;
-    }
+    Optional<ElectInstance> lastElect = getLastElectInstance(instanceId);
+    return lastElect.map(ElectInstance::getState).orElse(State.WAITING);
   }
 
   /**
-   * Add the given consensus to this node. If it was already present, do nothing.
+   * Add the given ElectInstance to this node. If it was already present, do nothing.
    *
-   * @param consensus the consensus to add
+   * @param electInstance the ElectInstance to add
    */
-  public void addConsensus(Consensus consensus) {
-    MessageID messageId = consensus.getMessageId();
-    // If the consensuses does not contain one linked this message id, add it
-    if (consensuses.stream().map(Consensus::getMessageId).noneMatch(messageId::equals)) {
-      consensuses.add(consensus);
-    }
+  public void addElectInstance(ElectInstance electInstance) {
+    electInstances.add(electInstance);
```

Comment:
Aren't you losing part of the check ?
It is now possible to add two ElectInstance with the same MessageID to the set.

Reply:
electInstances is now a Set and if the electInstance was already added, it won't be added twice.
ElectInstance do not override hashcode/equals, so it will be possible to add two ElectInstance only if it's two different object, but this shouldn't happens normally.
I can either revert the previous check, or I can override the equals of ElectInstance but this seems to be a bad idea because it is mutable.
Do you think it's better to only add the previous check ?